### PR TITLE
Alexey/features

### PR DESCRIPTION
**Pull request by alex-kovoy**
+ adding a feature support. Now initialization of Nodes and Sessions happens via feature activation. This is also used in E.
+ sessions now can be disabled/enabled based on RBAC.
+ refactoring of the webapi endpoints (userContext)

### CODE REVIEW

File: lib/web/apiserver.go [hunk 1]
```diff
@@ -1084,44 +1084,7 @@ type nodeWithSessions struct {
...
-            "server_addr": "127.0.0.1:3000",
-            "last_active": "time" // RFC3339 timestamp when user was last acive
-         }]
-     }]
-   }
-  ]
-}
-*/
-func (m *Handler) getSiteNodes(w http.ResponseWriter, r *http.Request, p httprouter.Params, c *SessionContext, site reversetunnel.RemoteSite) (interface{}, error) {
+func (m *Handler) siteNodesGet(w http.ResponseWriter, r *http.Request, p httprouter.Params, c *SessionContext, site reversetunnel.RemoteSite) (interface{}, error) {
```
Comment (russjones) [inline lib/web/apiserver.go]: Do we not use this endpoint to get sessions anymore?

File: lib/web/apiserver.go [hunk 2]
```diff
@@ -1135,30 +1098,13 @@ func (m *Handler) getSiteNodes(w http.ResponseWriter, r *http.Request, p httprou
...
-			}
-		}
-	}
-	nodes := make([]nodeWithSessions, 0, len(nodeMap))
-	for key := range nodeMap {
-		nodes = append(nodes, *nodeMap[key])
+
+	items := []interface{}{}
+	for _, server := range servers {
+		items = append(items, server.V1())
```
Comment (russjones) [inline lib/web/apiserver.go]: I think we have a `ServerSpecV2` and we should use the latest everywhere in code.

File: lib/web/ui/usercontext.go [hunk 3]
```diff
@@ -1,43 +1,91 @@
...
 		}
 	}
 
-	userACL := MergeAccessSet(accessSet)
+	return true
+}
+
+func newAccess(roleSet services.RoleSet, ctx *services.Context, kind string) access {
+	return access{
+		Read:   hasAccess(roleSet, ctx, kind, services.VerbList),
```
Comment (russjones) [inline lib/web/ui/usercontext.go]: Looks like you are checking `VerbList` and granting access to `Read` here, was this intentional?

File: lib/web/ui/usercontext.go [hunk 4]
```diff
@@ -1,43 +1,91 @@
...
+	"github.com/gravitational/teleport/lib/defaults"
 	"github.com/gravitational/teleport/lib/services"
-
-	"github.com/gravitational/trace"
+	"github.com/gravitational/teleport/lib/utils"
 )
 
+type access struct {
+	Read   bool `json:"read"`
+	Edit   bool `json:"edit"`
```
Comment (russjones) [inline lib/web/ui/usercontext.go]: Minor issue, but why not use `Update` here? Then it's consistent with the verbs on the backend.

File: lib/web/ui/usercontext.go [hunk 5]
```diff
@@ -1,43 +1,91 @@
...
-// NewUserContext returns userContext
-func NewUserContext(user services.User, allRoles []services.Role) (*userContext, error) {
-	userRoles := user.GetRoles()
-	roleNamesMap := map[string]bool{}
-	for _, name := range userRoles {
-		roleNamesMap[name] = true
+func getLogins(roleSet services.RoleSet) []string {
+	allLogins := []string{}
+	for _, role := range roleSet {
+		logins := role.GetLogins(services.Allow)
```
Comment (russjones) [inline lib/web/ui/usercontext.go]: Just a heads up, you are only requesting the allow logins, a login in an allow in one role could be a deny in another.

File: lib/web/apiserver.go (hunk 1 above)
Comment (alex-kovoy) [reply lib/web/apiserver.go]: Correct. We stopped using it a while back and did not clean the webapi endpoint.

File: lib/web/ui/usercontext.go (hunk 4 above)
Comment (alex-kovoy) [reply lib/web/ui/usercontext.go]: 'edit' brings a better semantic meaning to UI as it's mostly used for content editing. As for 'delete' - this is a reserved word in javascript.

File: lib/web/apiserver.go (hunk 2 above)
Comment (klizhentas) [reply lib/web/apiserver.go]: if possible, yes, I would like to grandfather the V1 spec of server, but not sure if it fits the scope of the change. may be create a separate issue/PR for this

File: lib/web/ui/usercontext.go [hunk 6]
```diff
@@ -1,43 +1,91 @@
...
+func newAccess(roleSet services.RoleSet, ctx *services.Context, kind string) access {
+	return access{
+		Read:   hasAccess(roleSet, ctx, kind, services.VerbList),
+		Edit:   hasAccess(roleSet, ctx, kind, services.VerbUpdate),
+		Create: hasAccess(roleSet, ctx, kind, services.VerbCreate),
+		Delete: hasAccess(roleSet, ctx, kind, services.VerbDelete),
+	}
+}
+
+// NewUserContext returns userContext
```
Comment (klizhentas) [inline lib/web/ui/usercontext.go]: this comment is not informative, I would rather change it to:

```NewUserContext constructs user context from roles assigned to user```

**Review by klizhentas — APPROVED** [review]

File: lib/web/ui/usercontext.go (hunk 3 above)
Comment (alex-kovoy) [reply lib/web/ui/usercontext.go]: Good catch! Will fix.

File: lib/web/apiserver.go (hunk 2 above)
Comment (alex-kovoy) [reply lib/web/apiserver.go]: It needs to be replaced with a new UI object (/lib/web/ui) which can be constructed using interfaces. I will create a separate issue for that.

File: lib/web/ui/usercontext.go [hunk 7]
```diff
@@ -1,43 +1,99 @@
...
-func NewUserContext(user services.User, allRoles []services.Role) (*userContext, error) {
-	userRoles := user.GetRoles()
-	roleNamesMap := map[string]bool{}
-	for _, name := range userRoles {
-		roleNamesMap[name] = true
+func getLogins(roleSet services.RoleSet) []string {
+	allowed := []string{}
+	denied := []string{}
+	for _, role := range roleSet {
+		denied = append(role.GetLogins(services.Deny), denied...)
```
Comment (russjones) [inline lib/web/ui/usercontext.go]: These should be:

```
denied = append(denied, role.GetLogins(services.Deny)...)
allowed = append(allowed, role.GetLogins(services.Allow)...)
```

At the moment you resetting the the slice at every iteration of the loop.

**Pull request merged**